File: Day #400/solution.js
// Prefix-sum array P (P[k]=sum of first k elems); sum(i,j)=P[j]-P[i]. Preprocess O(n), query O(1).
class RangeSum {
  constructor(L) {
    this.P = new Array(L.length + 1).fill(0);
    for (let k = 0; k < L.length; k++) this.P[k + 1] = this.P[k] + L[k];
  }
  sum(i, j) { return this.P[j] - this.P[i]; }
}

const L = [1, 2, 3, 4, 5];
const rs = new RangeSum(L);
console.log(rs.sum(1, 3));

File: Day #400/solution.cpp
// Prefix-sum array P (P[k]=sum of first k elems); sum(i,j)=P[j]-P[i]. Preprocess O(n), query O(1) time/space per query.
#include <iostream>
#include <vector>
using namespace std;

struct RangeSum {
    vector<long long> P;
    RangeSum(const vector<int>& L) {
        P.assign(L.size() + 1, 0);
        for (size_t k = 0; k < L.size(); ++k) P[k + 1] = P[k] + L[k];
    }
    long long sum(int i, int j) { return P[j] - P[i]; }
};

int main() {
    vector<int> L = {1, 2, 3, 4, 5};
    RangeSum rs(L);
    cout << rs.sum(1, 3) << "\n";
    return 0;
}
